File: javascript/src/bowling-game.js
var BowlingGame = function() {
  var rolls = [];

  var roll = function(pins){
    rolls.push(pins);
    if(pins == 10) rolls.push(0);
  };

  var score = function() {
    var score = 0;
    for (var frame = 1; frame <= 10; frame++){
      score += scoreFor(frame);
    }
    return score;
  };

  var scoreFor = function(frame){
    if (isStrike(frame)) return baseScoreFor(frame) + strikeBonusFor(frame);
    if  (isSpare(frame)) return baseScoreFor(frame) +  spareBonusFor(frame);
    else                 return baseScoreFor(frame);
  };

  var isStrike = function(frame){
    return firstRollOf(frame) == 10;
  };

  var strikeBonusFor = function(frame){
    return firstRollOf(frame + 1) + secondRollOf(frame + 1);
  };

  var isSpare = function(frame){
    return baseScoreFor(frame) == 10;
  };

  var spareBonusFor = function(frame){
    return firstRollOf(frame + 1);
  };

  function baseScoreFor(frame) {
    return firstRollOf(frame) + secondRollOf(frame);
  }

  var firstRollOf = function(frame){
    return rolls[(frame-1) * 2] || 0;
  };

  var secondRollOf = function(frame){
    return rolls[(frame-1) * 2 + 1] || 0;
  };

  return { roll: roll, score: score };
};

module.exports = BowlingGame;

File: ruby/lib/bowling_game.rb
class BowlingGame
  attr_reader :rolls
  def initialize
    @rolls = []
  end

  def roll *rolls
    rolls.each { |roll| @rolls << roll }
  end

  def score
    (1..10).reduce(0) do |score, frame|
      score + score_for(frame)
    end
  end

  def score_for frame
    return 10 + strike_bonus_for(frame) if strike? frame
    return 10 + spare_bonus_for(frame) if spare? frame
    first_roll_of(frame) + second_roll_of(frame)
  end

  def strike? frame
    first_roll_of(frame) == 10
  end

  def strike_bonus_for frame
    first_roll_of(frame + 1) + second_roll_of(frame + 1)
  end

  def spare? frame
    first_roll_of(frame) + second_roll_of(frame) == 10
  end

  def spare_bonus_for frame
    first_roll_of(frame + 1)
  end

  def first_roll_of frame
    @rolls[(frame-1)*2] || 0
  end

  def second_roll_of frame
    @rolls[(frame-1)*2 + 1] || 0
  end
end
